### PR TITLE
Added authClient.token.decode

### PR DESCRIPTION
**Pull request by lboyette-okta**
Resolves: OKTA-94712

@rchild-okta 
@stsai-okta

### CODE REVIEW

File: test/spec/token.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        done();
+      });
+    });
+
+    it('throws an error for a malformed token', function (done) {
+      return setup()
+      .then(function (oa) {
+        return oa.token.decode('malformedToken');
+      })
+      .then(function (res) {
```
Comment (stsai-okta) [inline test/spec/token.js]: is there something like a fail() in jasmine that will error out if it does hit this?

File: lib/token.js [hunk 2]
```diff
@@ -483,8 +487,8 @@ function getIdToken(sdk, oauthOptions, options) {
 }
 
 module.exports = {
-	getIdToken: getIdToken,
```
Comment (rchild-okta) [inline lib/token.js]: nice

File: test/spec/token.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+define(function(require) {
+  var OktaAuth = require('OktaAuth');
+  var Q = require('q');
+  var tokens = require('../util/tokens');
+
+  describe('token.decode', function () {
+
+    function setup() {
+      return Q.resolve(new OktaAuth({
```
Comment (rchild-okta) [inline test/spec/token.js]: What's the point of doing this? It's not async is it?

File: test/spec/token.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        url: 'http://example.okta.com'
+      }));
+    }
+
+    it('correctly decodes a token', function (done) {
+      return setup()
+      .then(function (oa) {
+        var decodedToken = oa.token.decode(tokens.unicodeToken);
```
Comment (rchild-okta) [inline test/spec/token.js]: Decoding is not an async action is it?

File: test/spec/token.js (hunk 1 above)
Comment (rchild-okta) [inline test/spec/token.js]: Could throw an error here that does not match the error message

Comment (lboyette-okta) [inline test/spec/token.js]: @stsai-okta unfortunately not

File: test/spec/token.js (hunk 4 above)
Comment (lboyette-okta) [inline test/spec/token.js]: It's not

File: test/spec/token.js (hunk 3 above)
Comment (lboyette-okta) [inline test/spec/token.js]: No, but we'll be adding more functionality to the token spec that will require promises. This is also the pattern we use in all our other specs.

File: test/spec/token.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+define(function(require) {
+  var OktaAuth = require('OktaAuth');
```
Comment (lboyette-okta) [inline test/spec/token.js]: They aren't yet. The core source is in commonjs, but the tests still use this amd style.

File: lib/token.js [hunk 6]
```diff
@@ -52,7 +52,11 @@ function decodeIdToken(idToken) {
       signature: jwt[2]
     };
   } catch(e) {
-    throw new AuthSdkError('Malformed idToken');
+    if (isIdToken) {
+      throw new AuthSdkError('Malformed idToken');
```
Comment (lboyette-okta) [inline lib/token.js]: This isn't something a user can do anything about, so the check should be on ANY error (maybe switching on `AuthSdkError` or `OAuthError`, but you're right that there shouldn't be a check on the string. I'll remove this logic here and change our tests to check for an error with a type of `AuthSdkError` and a non-empty errorSummary.

File: lib/clientBuilder.js [hunk 7]
```diff
@@ -69,13 +69,17 @@ function OktaAuthBuilder(args) {
     authorize: util.bind(token.getIdToken, sdk, sdk),
     verify: util.bind(token.verifyIdToken, sdk, sdk),
     refresh: util.bind(token.refreshIdToken, sdk, sdk),
-    decode: util.bind(token.decodeIdToken, sdk)
+    decode: util.bind(token.decodeToken, sdk, true)
```
Comment (lboyette-okta) [inline lib/clientBuilder.js]: alright

File: test/spec/token.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      var decodedToken = oa.token.decode(tokens.unicodeToken);
+      expect(decodedToken).toDeepEqual(tokens.unicodeDecoded);
+    });
+
+    it('throws an error for a malformed token', function () {
+      var oa = setupSync();
+      try {
+        var res = oa.token.decode('malformedToken');
+        // Should never hit this
+        expect(res).toBeNull();
```
Comment (rchild-okta) [inline test/spec/token.js]: nit: To be clearer, it's probably best to test something that we know will never be true, i.e. "expect(true).toBe(false)". In this case if the decode method does return null for this case, the test won't fail. Or use other suggestion of throwing an error that's not an AuthSdkError so it'll fail in the next block.

**Comment by rchild-okta** [comment]
Overall looks good, just modify that one test expectation and then 🚀

File: test/spec/token.js (hunk 8 above)
Comment (lboyette-okta) [inline test/spec/token.js]: fair enough. I'll change it

**Pull request merged**